### ISSUE TITLE
Variable.stack constructs extremely large chunks

### ISSUE THREAD

**Issue by dcherian** [pydata/xarray#5754]
**Minimal Complete Verifiable Example**:

Here's a small array with too-small chunk sizes just as an example
```python
# Put your MCVE code here
import dask.array
import xarray as xr

var = xr.Variable(("x", "y", "z"), dask.array.random.random((4, 18483, 1000), chunks=(1, 183, -1)))
```
<img width="400" alt="image" src="https://user-images.githubusercontent.com/2448579/131605110-53efbb3e-9d02-4ee7-8707-95c532671da6.png">

Now stack two dimensions, this is a 100x increase in chunk size (in my actual code, 85MB chunks become 8.5GB chunks =) )

```
var.stack(new=("x", "y"))
```
<img width="445" alt="image" src="https://user-images.githubusercontent.com/2448579/131605036-05aeab35-1e92-4b5e-ad17-30052a25743b.png">

But calling `reshape` on the dask array preserves the original chunk size
``` 
var.data.reshape((4*18483, -1))
```
<img width="388" alt="image" src="https://user-images.githubusercontent.com/2448579/131605076-8ffbdde9-c8b2-48e7-9acd-00bb8309436d.png">

## Solution

Ah, found it , we transpose then reshape in `Variable_stack_once`. 
https://github.com/pydata/xarray/blob/f915515d610b4471888fa44dfb00dbae3fd22349/xarray/core/variable.py#L1521-L1527

Writing those steps with pure dask yields the same 100x increase in chunksize

``` python
var.data.transpose([2, 0, 1]).reshape((-1, 4*18483))
```
<img width="398" alt="image" src="https://user-images.githubusercontent.com/2448579/131605694-a3bb8859-df4a-49a7-83cd-d16bb203e77e.png">

**Anything else we need to know?**:

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-1127.18.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.19.0
pandas: 1.3.1
numpy: 1.21.1
scipy: 1.5.3
netCDF4: 1.5.6
pydap: installed
h5netcdf: 0.11.0
h5py: 3.3.0
Nio: None
zarr: 2.8.3
cftime: 1.5.0
nc_time_axis: 1.3.1
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: 3.0.4
bottleneck: 1.3.2
dask: 2021.07.2
distributed: 2021.07.2
matplotlib: 3.4.2
cartopy: 0.19.0.post1
seaborn: 0.11.1
numbagg: None
pint: 0.17
setuptools: 49.6.0.post20210108
pip: 21.2.2
conda: 4.10.3
pytest: 6.2.4
IPython: 7.26.0
sphinx: 4.1.2
</details>

**Comment by dcherian** [comment]
Ah this is https://github.com/dask/dask/issues/5544 again. It looks like dask needs to break up the potentially-very-large intermediate chunks

**Comment by dcherian** [comment]
Fixed upstream

**Comment by yucsong** [comment]
Sorry, is this fixed?

**Comment by yucsong** [comment]
/srv/conda/envs/notebook/lib/python3.10/site-packages/xarray/core/variable.py:1721: PerformanceWarning: Reshaping is producing a large chunk. To accept the large
chunk and silence this warning, set the option
    >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):
    ...     array.reshape(shape)

I simply tested var.stack(new=("x", "y")) and I got the above message.   I don't understand why From 1521 to 1527 line of xarray/core/variable.py they did reshape?
